### PR TITLE
Support non-tuple expression for exists-subquery to join

### LINKED ISSUE #4366: Failed to create Left anti join physical plan due to SchemaError::FieldNotFound
**Describe the bug**

````
        let ctx = create_join_context_with_null_ids("t1_id", "t2_id", repartition_joins).unwrap();
        let sql = "SELECT t1_id, t1_name FROM t1 WHERE NOT EXISTS (SELECT 1 FROM t2 WHERE t1_id = t2_id and t1_id > 11) ORDER BY t1_id";
        let actual = execute_to_batches(&ctx, sql).await;
        let expected = vec![
            "+-------+---------+",
            "| t1_id | t1_name |",
            "+-------+---------+",
            "| 11    | a       |",
            "| 11    | a       |",
            "| 33    | c       |",
            "|       | e       |",
            "+-------+---------+",
        ];
        assert_batches_eq!(expected, &actual);

````

called `Result::unwrap()` on an `Err` value: "SchemaError(FieldNotFound { field: Column { relation: Some(\"t1\"), name: \"t1_id\" }, valid_fields: Some([Column { relation: Some(\"t2\"), name: \"t2_id\" }]) }) at Creating physical plan for 'SELECT t1_id, t1_name FROM t1 WHERE NOT EXISTS (SELECT 1 FROM t2 WHERE t1_id = t2_id and t1_id > 11) ORDER BY t1_id': Sort: t1.t1_id ASC NULLS LAST\n  Projection: t1.t1_id, t1.t1_name\n    LeftAnti Join: t1.t1_id = t2.t2_id\n      TableScan: t1 projection=[t1_id, t1_name]\n      Filter: CAST(t1.t1_id AS Int64) > Int64(11)\n        TableScan: t2 projection=[t2_id]"

**To Reproduce**

**Expected behavior**

**Additional context**

### LINKED ISSUE #4934: Support non-tuple expression for Exists Subquery to Join
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
(This section helps Arrow developers understand the context and *why* for this feature, in addition to  the *what*)

[PR-4826](https://github.com/apache/arrow-datafusion/pull/4826) had already added the support for in-subquery to join, but the exists subquery is left unchanged and the below SQL still does not work:

```sql
select t1.t1_id, t1.t1_name, t1.t1_int from t1 where EXISTS (select 1 from t2 where t2.t2_id + 1 = t1.t1_id + 12)
```

```rust
NotImplemented("Physical plan does not support logical expression EXISTS (<subquery>)")
```

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by ygf11**
# Which issue does this PR close?

Closes #4934.
Closes #4366.

# Rationale for this change

This sql works correctly in datafusion, and will be rewrited to `LeftSemi Join`. 
```sql
> explain select * from t1 where exists (select 1 from t2 where t1.t1_id = t2.t2_id);
+---------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| logical_plan  | Projection: t1.t1_id, t1.t1_name, t1.t1_int                                                                                                                      |
|               |   LeftSemi Join: t1.t1_id = t2.t2_id                                                                                                                             |
|               |     TableScan: t1 projection=[t1_id, t1_name, t1_int]                                                                                                            |
|               |     TableScan: t2 projection=[t2_id]    
```

But the following sql will not:
```sql
> select * from t1 where exists (select 1 from t2 where t1.t1_id + 1 = t2.t2_id * 2);
```

These sql should also be rewrited to `LeftSemi Join`.

# What changes are included in this PR?

# Are these changes tested?
Yes.

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/optimizer/src/decorrelate_where_exists.rs [hunk 1]
```diff
@@ -241,13 +252,14 @@ mod tests {
...
-    TableScan: orders [o_orderkey:Int64, o_custkey:Int64, o_orderstatus:Utf8, o_totalprice:Float64;N]"#;
-
+        let expected = "Projection: customer.c_custkey [c_custkey:Int64]\
+                        \n  LeftSemi Join:  Filter: orders.o_custkey = customer.c_custkey [c_custkey:Int64, c_name:Utf8]\
+                        \n    LeftSemi Join:  Filter: orders.o_custkey = customer.c_custkey [c_custkey:Int64, c_name:Utf8]\
+                        \n      TableScan: customer [c_custkey:Int64, c_name:Utf8]\
+                        \n      Projection: orders.o_custkey [o_custkey:Int64]\
+                        \n        TableScan: orders [o_orderkey:Int64, o_custkey:Int64, o_orderstatus:Utf8, o_totalprice:Float64;N]\
+                        \n    Projection: orders.o_custkey [o_custkey:Int64]\
+                        \n      TableScan: orders [o_orderkey:Int64, o_custkey:Int64, o_orderstatus:Utf8, o_totalprice:Float64;N]";
```
Comment (ygf11) [inline datafusion/optimizer/src/decorrelate_where_exists.rs:262]: Before this pr, this rule will not keep the projection.
I think keeping projection make it easy to read, and spark also keeps it.

File: datafusion/optimizer/src/decorrelate_where_exists.rs [hunk 2]
```diff
@@ -305,21 +318,21 @@ mod tests {
...
     }
 
-    /// Test for correlated exists subquery with no columns in schema
     #[test]
     fn exists_subquery_no_cols() -> Result<()> {
         let sq = Arc::new(
             LogicalPlanBuilder::from(scan_tpch_table("orders"))
-                .filter(col("customer.c_custkey").eq(col("customer.c_custkey")))?
+                .filter(col("customer.c_custkey").eq(lit(1u32)))?
                 .project(vec![col("orders.o_custkey")])?
```
Comment (ygf11) [inline datafusion/optimizer/src/decorrelate_where_exists.rs:336]: `customer.c_custkey = customer.c_custkey` is too specific here, so modify it to `customer.c_custkey = 1`.

File: benchmarks/expected-plans/q21.txt [hunk 3]
```diff
@@ -14,8 +14,10 @@ Sort: numwait DESC NULLS FIRST, supplier.s_name ASC NULLS LAST
...
-          Filter: lineitem.l_receiptdate > lineitem.l_commitdate
-            TableScan: lineitem projection=[l_orderkey, l_suppkey, l_commitdate, l_receiptdate]
\ No newline at end of file
+          Projection: l2.l_orderkey, l2.l_suppkey
+            SubqueryAlias: l2
+              TableScan: lineitem projection=[l_orderkey, l_suppkey]
+        Projection: l3.l_orderkey, l3.l_suppkey
+          SubqueryAlias: l3
+            Filter: lineitem.l_receiptdate > lineitem.l_commitdate
+              TableScan: lineitem projection=[l_orderkey, l_suppkey, l_commitdate, l_receiptdate]
```
Comment (ygf11) [inline benchmarks/expected-plans/q21.txt:23]: The difference is we keep the projection now.

File: datafusion/core/tests/sql/joins.rs [hunk 4]
```diff
@@ -2187,7 +2187,6 @@ async fn left_anti_join() -> Result<()> {
 }
 
 #[tokio::test]
-#[ignore = "Test ignored, will be enabled after fixing the anti join plan bug"]
 // https://github.com/apache/arrow-datafusion/issues/4366
```
Comment (ygf11) [inline datafusion/core/tests/sql/joins.rs:2190]: This test passes now

**Comment by alamb** [comment]
I will try and find time to review this PR over the next day or two

File: datafusion/optimizer/src/decorrelate_where_exists.rs [hunk 5]
```diff
@@ -144,55 +142,68 @@ fn optimize_exists(
...
-    let join_keys = (subqry_cols, outer_cols);
+    let join_filter = conjunction(join_filters);
 
     // join our sub query into the main plan
     let join_type = match query_info.negated {
         true => JoinType::LeftAnti,
         false => JoinType::LeftSemi,
     };
+
+    // TODO: add Distinct if the original plan is a Distinct.
```
Comment (alamb) [inline datafusion/optimizer/src/decorrelate_where_exists.rs:197]: is this still a todo? I see the code above looking into Distinct children, but the distinct is not added back

**Review by alamb — COMMENTED** [review]
Thank you @ygf11 

 I went over this code and the plan changes carefully and they look great to me.

The only thing I think needs to be done prior is double check handling of distincts. Otherwise really nice work. Thank you 🏆

File: datafusion/optimizer/src/utils.rs [hunk 6]
```diff
@@ -468,6 +469,42 @@ pub fn merge_schema(inputs: Vec<&LogicalPlan>) -> DFSchema {
...
+    if let LogicalPlan::Filter(plan_filter) = maybe_filter {
+        let input_schema = plan_filter.input.schema();
+        let subquery_filter_exprs = split_conjunction(&plan_filter.predicate);
+
+        let mut join_filters: Vec<Expr> = vec![];
+        let mut subquery_filters: Vec<Expr> = vec![];
+        for expr in subquery_filter_exprs {
+            let cols = expr.to_columns()?;
+            if check_all_column_from_schema(&cols, input_schema.clone()) {
+                subquery_filters.push(expr.clone());
```
[the remark this answers is not in the document]
Comment (ygf11) [reply datafusion/optimizer/src/utils.rs:490]: I think we need clone here.

`subquery_filter_exprs` is the result of `split_conjunction` function, and the type is `Vec<&Expr>`.

https://github.com/apache/arrow-datafusion/blob/fed4019d556f4afb3156fd12c21608e08b8d7eb6/datafusion/optimizer/src/utils.rs#L65-L67

File: datafusion/optimizer/src/decorrelate_where_exists.rs (hunk 5 above)
Comment (ygf11) [reply datafusion/optimizer/src/decorrelate_where_exists.rs:197]: Yes, it is still a todo, some distinct cases need more checks. like:
```rust
// SELECT t1.t1_id,
//       t1.t1_name,
//       t1.t1_int
// FROM   t1
// WHERE  EXISTS(SELECT DISTINCT t2_int
//              FROM   t2
//              WHERE  t2.t2_id > t1.t1_id); 

// if we just add back the `DISTINCT`, the result:
Projection: t1.t1_id, t1.t1_name, t1.t1_int
  LeftSemi Join:  Filter: t2.t2_int > t1.t1_int
    TableScan: t1
    Distinct:
      Projection: t2.t2_int
         TableScan: t2
 
// expected result:
Projection: t1.t1_id, t1.t1_name, t1.t1_int
  LeftSemi Join:  Filter: t2.t2_int > t1.t1_int
    TableScan: t1
    Distinct:
      Projection: t2.t2_int, t2.t2_id
         TableScan: t2
```

`t2_int` will not be in the projection.

The reason is we just consider the columns from join filter as the project items, we should also consider columns from original projection when there is an outer distinct.

I will do this in the following pr if it is ok :rofl:.

Comment (alamb) [reply datafusion/optimizer/src/decorrelate_where_exists.rs:197]: > I will do this in the following pr if it is ok 🤣.

Absolutely it is ok! Thank you for all your work

**Review by alamb — APPROVED** [review]
Thanks again @ygf11 !

There appears to be a conflict in this PR -- I think once that is fixed we can merge it in.

**Comment by ygf11** [comment]
Thanks for reviewing  @alamb. I fixed the conflict.

File: datafusion/core/tests/sql/joins.rs (hunk 4 above)
Comment (ygf11) [reply datafusion/core/tests/sql/joins.rs]: Linked issue #4366 to this pr.

**Comment by mingmwang** [comment]
@ygf11 @jackwener 
I have a feeling that maybe we should combine the two rules `DecorrelateWhereExists` and `DecorrelateWhereIn` to one rule. How do you think ?

**Comment by ygf11** [comment]
> I have a feeling that maybe we should combine the two rules DecorrelateWhereExists and DecorrelateWhereIn to one rule. How do you think ?

Agree, most logics are same, we should combine them. I can take this task if others do not work on  :smile:.

**Comment by mingmwang** [comment]
One more thing, looks like the SubqueryAlias logic is not consistent between the two rules `DecorrelateWhereExists` and `DecorrelateWhereIn`
https://github.com/apache/arrow-datafusion/pull/4767

I think you can make them consistent in the refactoring work.

```rust
 let sql = "SELECT t1_id, t1_name FROM t1 WHERE t1_id IN (SELECT t2_id FROM t2) ORDER BY t1_id";
        let msg = format!("Creating logical plan for '{sql}'");
        let dataframe = ctx.sql(&("explain ".to_owned() + sql)).await.expect(&msg);
        let plan = dataframe.into_optimized_plan()?;
        let expected = vec![
            "Explain [plan_type:Utf8, plan:Utf8]",
            "  Sort: t1.t1_id ASC NULLS LAST [t1_id:UInt32;N, t1_name:Utf8;N]",
            "    Projection: t1.t1_id, t1.t1_name [t1_id:UInt32;N, t1_name:Utf8;N]",
            "      LeftSemi Join: t1.t1_id = __correlated_sq_1.t2_id [t1_id:UInt32;N, t1_name:Utf8;N]",
            "        TableScan: t1 projection=[t1_id, t1_name] [t1_id:UInt32;N, t1_name:Utf8;N]",
            "        SubqueryAlias: __correlated_sq_1 [t2_id:UInt32;N]",
            "          Projection: t2.t2_id AS t2_id [t2_id:UInt32;N]",
            "            TableScan: t2 projection=[t2_id] [t2_id:UInt32;N]",
        ];

```

```rust
let sql = "SELECT t1_id, t1_name FROM t1 WHERE NOT EXISTS (SELECT 1 FROM t2 WHERE t1_id = t2_id and t1_id > 11) ORDER BY t1_id";
        let msg = format!("Creating logical plan for '{sql}'");
        let dataframe = ctx.sql(&("explain ".to_owned() + sql)).await.expect(&msg);
        let plan = dataframe.into_optimized_plan()?;
        let expected = vec![
            "Explain [plan_type:Utf8, plan:Utf8]",
            "  Sort: t1.t1_id ASC NULLS LAST [t1_id:UInt32;N, t1_name:Utf8;N]",
            "    Projection: t1.t1_id, t1.t1_name [t1_id:UInt32;N, t1_name:Utf8;N]",
            "      LeftAnti Join: t1.t1_id = t2.t2_id Filter: t1.t1_id > UInt32(11) [t1_id:UInt32;N, t1_name:Utf8;N]",
            "        TableScan: t1 projection=[t1_id, t1_name] [t1_id:UInt32;N, t1_name:Utf8;N]",
            "        Projection: t2.t2_id [t2_id:UInt32;N]",
            "          TableScan: t2 projection=[t2_id] [t2_id:UInt32;N]",
        ];
```

**Comment by mingmwang** [comment]
BTW,  this PR looks really nice and fixed couple of issues related to the Exist subquery. 
Thanks you @ygf11.

**Comment by jackwener** [comment]
> @ygf11 @jackwener I have a feeling that maybe we should combine the two rules `DecorrelateWhereExists` and `DecorrelateWhereIn` to one rule. How do you think ?

Agree it.

**Pull request merged**